_posts/2019-01-09-a-new-recovery-process-for-real-time-pipelines/index.md

```diff
@@ -0,0 +1,302 @@
+    {
+      "name": "ReplaceInBase64FieldInBody",
+      "error": "instance type (string) does not match any allowed primitive type (allowed: [\"integer\"])\n    level: \"error\"\n    schema: {\"loadingURI\":\"#\",\"pointer\":\"/properties/sessionIndex\"",
+      "base64Field": "ue_px",
+      "toReplace": "\"sessionIndex\":\"(\\d+)\"",
+      # $1 refers to the first capture group
+      "replacement": "\"sessionIndex\":$1"
+    },
+    # Our custom recovery scenario, replaces a wrong Iglu webhook path
+    {
+      "name": "ReplaceInPath",
+      "error": "Payload with vendor com.iglu and version v1 not supported",
+      "toReplace": "com.iglu/v1",
+      "replacement": "com.snowplowanalytics.iglu/v1"
+    }
+  ]
+}
+{% endhighlight %}
+
+<h3 id="config">2.4 Testing</h3>
+
+It's possible to test an entire recovery without running it or a custom recovery scenario by
+following [the dedicated guide in our repository][recovery-testing].
+
+<h2 id="aws">3. Snowplow Event Recovery on AWS</h2>
+
+For AWS users, the recovery will take the form of a Spark job which you can run through EMR, for
+example. It will bad rows from an S3 location, run the recovery on this data, and store the
+recovered payloads in another S3 location.
+
+You can run the job using the JAR directly (which is hosted at
+`s3://snowplow-hosted-assets/3-enrich/snowplow-event-recovery/`):
+
+{% highlight bash %}
+spark-submit \
+  --class com.snowplowanalytcs.snowplow.event.recovery.Main \
+  --master master-url \
+  --deploy-mode deploy-mode \
+  snowplow-event-recovery-spark-0.1.0.jar
+  --input s3://bad-rows-location/
+  --output s3://recovered-collector-payloads-location/
+  --config base64-encoded-configuration
+{% endhighlight %}
+
+Or through an EMR step:
+
+{% highlight bash %}
+aws emr add-steps --cluster-id j-XXXXXXXX --steps \
+  Name=snowplow-event-recovery,\
+  Type=CUSTOM_JAR,\
+  Jar=s3://snowplow-hosted-assets/3-enrich/snowplow-event-recovery/snowplow-event-recovery-spark-0.1.0.jar,\
+  MainClass=com.snowplowanalytics.snowplow.event.recovery.Main,\
+  Args=[--input,s3://bad-rows-location/,--output,s3://recovered-collector-payloads-location/,--config,base64-encoded-configuration],\
+  ActionOnFailure=CONTINUE
+{% endhighlight %}
+
+Note that the configuration discussed above will need to be base64-encoded.
+
+<h2 id="gcp">4. Snowplow Event Recovery on GCP</h2>
+
+For GCP users, leveraging the data outputted by [the Snowplow Google Cloud Storage Loader][sgcsl],
+recovery will take the shape of a Beam job runnable on Dataflow. It will read bad rows from a GCS
+location specified through a pattern, run the recovery on this data, and store the the recovered
+payloads in a PubSub topic (ideally your PubSub topic containing the raw payloads so that fixed
+payloads can be directly picked up by the enrichment process).
+
+You can run the job using the zip archive, which can be downloaded from Bintray
+[here][bintray-archive]:
+
```

Comment:
Please add this link to the end of the post!

Reply:
addressed in https://github.com/snowplow/snowplowanalytics.com/pull/1027/commits/a06e8769cb76c0054fd283d8418bc8ad963c1fbd